src/meta/processors/partsMan/CreateSpaceProcessor.cpp

```diff
@@ -174,34 +200,97 @@ CreateSpaceProcessor::pickHosts(PartitionID partId,
     return pickedHosts;
 }
 
-std::vector<HostAddr>
-CreateSpaceProcessor::pickHostsWithZone(PartitionID partId,
-                                        const std::vector<std::string>& zones,
-                                        int32_t replicaFactor) {
-    auto startIndex = partId;
-    std::vector<HostAddr> pickedHosts;
-    for (int32_t i = 0; i < replicaFactor; i++) {
-        auto zoneName = zones[startIndex++ % zones.size()];
-        auto zoneIdRet = getZoneId(zoneName);
-        if (!zoneIdRet.ok()) {
-            LOG(ERROR) << "Zone " << zoneName << " not found";
-            handleErrorCode(cpp2::ErrorCode::E_NOT_FOUND);
-            onFinished();
-            return pickedHosts;
+StatusOr<std::unordered_map<HostAddr, int32_t>>
+CreateSpaceProcessor::getHostLoading() {
+    std::unique_ptr<kvstore::KVIterator> iter;
+    const auto& prefix = MetaServiceUtils::partPrefix();
+    auto code = kvstore_->prefix(kDefaultSpaceId, kDefaultPartId, prefix, &iter);
+    if (code != kvstore::ResultCode::SUCCEEDED) {
+        LOG(ERROR) << "List Parts Failed";
+        return Status::Error("List Parts Failed");
+    }
+
+    std::unordered_map<HostAddr, int32_t> result;
+    while (iter->valid()) {
+        auto hosts = MetaServiceUtils::parsePartVal(iter->val());
+        for (auto& host : hosts) {
+            result[host]++;
         }
+        iter->next();
+    }
+    return result;
+}
 
+StatusOr<std::vector<HostAddr>>
+CreateSpaceProcessor::pickHostsWithZone(const std::vector<std::string>& zones,
+                                        std::unordered_map<HostAddr, int32_t>& loading) {
+    std::vector<HostAddr> pickedHosts;
+    for (auto& zoneName : zones) {
         auto zoneKey = MetaServiceUtils::zoneKey(zoneName);
         auto zoneValueRet = doGet(std::move(zoneKey));
+        auto hosts = MetaServiceUtils::parseZoneHosts(std::move(zoneValueRet).value());
+
+        HostAddr picked;
+        int32_t size = INT_MAX;
+        for (auto& host : hosts) {
+            auto hostIter = loading.find(host);
+            if (hostIter == loading.end()) {
+                LOG(ERROR) << "Host " << host << " not found";
+                handleErrorCode(cpp2::ErrorCode::E_NO_HOSTS);
+                onFinished();
+                return Status::Error("Host not found");
+            }
+
+            if (size > hostIter->second) {
+                picked = host;
+                size = hostIter->second;
+            }
+        }
+
+        loading[picked] += 1;
+        pickedHosts.emplace_back(toThriftHost(std::move(picked)));
+    }
+    return pickedHosts;
+}
+
+StatusOr<std::vector<std::string>>
+CreateSpaceProcessor::pickLightLoadZones(const std::vector<std::string>& zones,
```

Comment:
IMO, one way to do it is use priority_queue, for each partition, just dequeue read first `replica` zones, then enqueue again.

Reply:
I have thought about heap (priority_queue) but the zone size is very small and use sort maybe more readable

Reply:
Yeah, the pros is that heap could be use several times, no need to get zones loading every time when add a new partition.